ui/v2.5/src/models/list-filter/criteria/criterion.ts

```diff
@@ -257,47 +119,69 @@ export abstract class Criterion {
       valueString = this.getLabelValue();
     }
 
-    return `${Criterion.getLabel(this.type)} ${modifierString} ${valueString}`;
+    return `${intl.formatMessage({
+      id: this.criterionOption.messageID,
+    })} ${modifierString} ${valueString}`;
   }
 
   public getId(): string {
-    return `${this.parameterName}-${this.modifier.toString()}`; // TODO add values?
+    return `${this.criterionOption.parameterName}-${this.modifier.toString()}`; // TODO add values?
   }
 
-  private static replaceSpecialCharacter(str: string, c: string) {
-    return str.replaceAll(c, encodeURIComponent(c));
+  public encodeValue(): V {
+    return this.value;
   }
 
-  public encodeValue(): CriterionValue {
-    // replace certain characters
-    if (typeof this.value === "string") {
-      let ret = this.value;
-      ret = Criterion.replaceSpecialCharacter(ret, "&");
-      ret = Criterion.replaceSpecialCharacter(ret, "+");
-      return ret;
-    }
-    return this.value;
+  public toJSON() {
+    const encodedCriterion = {
+      type: this.criterionOption.value,
+      // #394 - the presence of a # symbol results in the query URL being
+      // malformed. We could set encode: true in the queryString.stringify
+      // call below, but this results in a URL that gets pretty long and ugly.
+      // Instead, we'll encode the criteria values.
+      value: this.encodeValue(),
+      modifier: this.modifier,
+    };
+    return JSON.stringify(encodedCriterion);
   }
-}
 
-export interface ICriterionOption {
-  label: string;
-  value: CriterionType;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  public apply(outputFilter: Record<string, any>) {
+    // eslint-disable-next-line no-param-reassign
+    outputFilter[this.criterionOption.parameterName] = this.toCriterionInput();
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  protected toCriterionInput(): any {
+    return {
+      value: this.value,
+      modifier: this.modifier,
+    };
+  }
 }
 
-export class CriterionOption implements ICriterionOption {
-  public label: string;
-  public value: CriterionType;
+// used in various places, to eliminate the need to disable eslint rules
+// every time its used.
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type CriterionAny = Criterion<any>;
```

Comment:
You can use `Criterion<CriterionValue>` rather than having to use `any`.

Reply:
Awesome, thanks.